### PR TITLE
Advise options reuse

### LINKED ISSUE #19878: Advise use of a single options instance in a custom converter
See https://github.com/dotnet/runtime/issues/40072 and https://github.com/dotnet/runtime/issues/42167#issuecomment-691700983

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5de37d79-751f-e372-e866-efde28f4908c
* Version Independent ID: 678c4f23-764a-7ccd-0d6d-159a30957477
* Content: [How to write custom converters for JSON serialization - .NET](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to)
* Content Source: [docs/standard/serialization/system-text-json-converters-how-to.md](https://github.com/dotnet/docs/blob/master/docs/standard/serialization/system-text-json-converters-how-to.md)
* Product: **dotnet**
* GitHub Login: @tdykstra
* Microsoft Alias: **tdykstra**

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #19878

[Internal review URL](https://review.docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-configure-options?branch=pr-en-us-21791)

### CODE REVIEW

File: docs/standard/serialization/snippets/system-text-json-configure-options/csharp/SystemTextJsonOptions.csproj [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net5.0</TargetFramework>
```
Comment (IEvangelist) [inline docs/standard/serialization/snippets/system-text-json-configure-options/csharp/SystemTextJsonOptions.csproj:5]:
```suggestion
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <Nullable>enabled</Nullable>
```

File: docs/standard/serialization/snippets/system-text-json-configure-options/csharp/ReuseOptionsInstances.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            if (options == null)
+            {
+                options = new JsonSerializerOptions()
+                {
+                    WriteIndented = true
+                };
+            }
+
+            _ = JsonSerializer.Serialize<Forecast>(forecast, options);
+        }
```
Comment (IEvangelist) [inline docs/standard/serialization/snippets/system-text-json-configure-options/csharp/ReuseOptionsInstances.cs:45]:
```suggestion
        private static void Serialize(Forecast forecast, JsonSerializerOptions? options)
        {
            _ = JsonSerializer.Serialize<Forecast>(
                forecast,
                options ?? = new JsonSerializerOptions()
                {
                    WriteIndented = true
                });
        }
```

**Review by IEvangelist — APPROVED** [review]
Minor suggestions for C# bits - otherwise, looks good to me, :shipit: when you're ready.

File: docs/standard/serialization/snippets/system-text-json-configure-options/csharp/ReuseOptionsInstances.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            watch = Stopwatch.StartNew();
+            for (int i = 0; i < iterations; i++)
+            {
+                Serialize(forecast, null);
+            }
+            watch.Stop();
+            Console.WriteLine($"Elapsed time creating new options instances: {watch.ElapsedMilliseconds}");
+        }
+
+        private static void Serialize(Forecast forecast, JsonSerializerOptions? options)
```
Comment (layomia) [inline docs/standard/serialization/snippets/system-text-json-configure-options/csharp/ReuseOptionsInstances.cs:34]:
```suggestion
        private static void Serialize(Forecast forecast, JsonSerializerOptions? options = null)
```

File: docs/standard/serialization/snippets/system-text-json-configure-options/csharp/ReuseOptionsInstances.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                Serialize(forecast, options);
+            }
+            watch.Stop();
+            Console.WriteLine($"Elapsed time using one options instance: {watch.ElapsedMilliseconds}");
+
+            watch = Stopwatch.StartNew();
+            for (int i = 0; i < iterations; i++)
+            {
+                Serialize(forecast, null);
```
Comment (layomia) [inline docs/standard/serialization/snippets/system-text-json-configure-options/csharp/ReuseOptionsInstances.cs:28]:
```suggestion
                Serialize(forecast);
```

File: docs/standard/serialization/system-text-json-configure-options.md [hunk 5]
```diff
@@ -13,7 +13,21 @@ helpviewer_keywords:
...
+
+The preceding code serializes a small object 100,000 times using the same options instance. Then it serializes the same object the same number of times and creates a new options instance each time. A typical run time difference is 190 compared to 40,140 milliseconds. The difference is even greater if you increase the number of iterations.
+
+The serializer undergoes a warm-up phase during the first serialization of each type in the object graph when a new options instance is passed to it. This warm-up includes creating a cache of metadata that is needed for serialization. The metadata includes delegates to property getters, setters, constructor arguments, specified attributes, and so forth. This metadata cache is stored in the options instance. The same warm-up process and cache applies to deserialization.
+
+The size of the metadata cache in a `JsonSerializerOptions` instance depends on the number of types to be serialized. If you pass numerous dynamically generated types to the serializer, the cache size will continue to grow and can end up causing an `OutOfMemoryException`.
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-configure-options.md:30]: Dynamic type generation is not the only way to have numerous types (although it is the most likely)

```suggestion
The size of the metadata cache in a `JsonSerializerOptions` instance depends on the number of types to be serialized. If you pass numerous types—for example, dynamically generated types—to the serializer, the cache size will continue to grow and can end up causing an `OutOfMemoryException`.
```

(Or something like this...)

**Review by layomia — APPROVED** [review]
Very nice, thanks. We may need to augment the guidance following work in https://github.com/dotnet/runtime/issues/40072#issuecomment-670646253, but this is very useful given the current behavior.

**Pull request merged**